packages/grid/x-data-grid/src/hooks/features/columnGrouping/useGridColumnGroupingPreProcessors.ts

```diff
@@ -0,0 +1,41 @@
+import * as React from 'react';
+import { GridPipeProcessor, useGridRegisterPipeProcessor } from '../../core/pipeProcessing';
+import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
+import { GridApiCommunity } from '../../../models/api/gridApiCommunity';
+import { isDeepEqual } from '../../../utils/utils';
+import { unwrapGroupingColumnModel, hasGroupPath } from './useGridColumnGrouping';
+
+export const useGridColumnGroupingPreProcessors = (
+  apiRef: React.MutableRefObject<GridApiCommunity>,
+  props: DataGridProcessedProps,
+) => {
+  const addHeaderGroups = React.useCallback<GridPipeProcessor<'hydrateColumns'>>(
+    (columnsState) => {
+      if (!props.experimentalFeatures?.columnGrouping || !props.columnGroupingModel) {
```

Comment:
Should grid react to `columnGroupingModel` change?
If so, then preprocessing should run even if `columnGroupingModel` is `undefined` and, most importantly, make sure that those groups are no longer in the state.

Here's a demo to demonstrate the problem: https://codesandbox.io/s/basicgroupingdemo-demo-mui-x-forked-jm6ynq?file=/demo.tsx
Try to enable grouping and then disable it